cmd/gator/validate/validate.go

```
@@ -0,0 +1,200 @@
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	if stdinfo.Size() > 0 {
+		us, err := readYAMLSource(os.Stdin)
+		if err != nil {
+			errLog.Fatalf("reading from stdin: %s", err)
+		}
+		unstrucs = append(unstrucs, us...)
+	} else if len(flagFilenames) > 0 {
+		// normalize directories by listing their files
+		normalized, err := normalize(flagFilenames)
+		if err != nil {
+			errLog.Fatalf("normalizing: %s", err)
+		}
+
+		for _, filename := range normalized {
+			file, err := os.Open(filename)
+			if err != nil {
+				errLog.Fatalf("opening file %q: %s", filename, err)
+			}
+
+			us, err := readYAMLSource(bufio.NewReader(file))
+			if err != nil {
+				errLog.Fatalf("reading from file %q: %s", filename, err)
+			}
+			file.Close()
+
+			unstrucs = append(unstrucs, us...)
+		}
+	} else {
+		errLog.Fatalf("no input data: must include data via either stdin or the %q flag", flagNameFilename)
+	}
+
+	responses, err := validate.Validate(unstrucs)
+	if err != nil {
+		errLog.Fatalf("auditing objects: %v\n", err)
+	}
+
+	results := responses.Results()
+
+	if flagYAML && flagJSON {
+		errLog.Fatalf("cannot use --%s and --%s in combination", flagNameJSON, flagNameYAML)
+	}
+
+	if flagJSON {
+		b, err := json.MarshalIndent(results, "", "    ")
+		if err != nil {
+			errLog.Fatalf("marshaling validation json results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if flagYAML {
+		b, err := yaml.Marshal(results)
+		if err != nil {
+			errLog.Fatalf("marshaling validation yaml results: %s", err)
+		}
+		outLog.Fatal(string(b))
+	}
+
+	if len(results) > 0 {
+		for _, result := range results {
+			outLog.Printf("Message: %q", result.Msg)
+		}
+
+		os.Exit(1)
+	}
+}
+
+func normalize(filenames []string) ([]string, error) {
+	var output []string
+
+	for _, filename := range filenames {
+		err := filepath.Walk(filename, func(path string, info os.FileInfo, err error) error {
```

What happens if we run into a non-YAML file?

---

One possible option is try to ingest all files and silently skip if they can't be ingested.  This would take care of non yaml/json files.  The downside is it would swallow errors due to common mistakes like malformed config.

Another option is to check file extension.  `kubectl` supports `[.json .yaml .yml]`, so we can do the same.  `kubectl` also supports a file with no extension if you specify it directly with `kubectl apply -f file-without-extension`.  I can do that as well.

---

Now supports `[.json, .yaml, .yml]` and skips everything else.

---

I wrote a test and it turns out that these extensions are supported OOTB by the `YAMLToJSONDecoder` I used.  Thanks apimachinery!